src/Client/JWTProvider.cpp

```diff
@@ -0,0 +1,272 @@
+#include <config.h>
+
+#if USE_JWT_CPP && USE_SSL
+#include <Client/JWTProvider.h>
+#include <Common/Exception.h>
+
+#include <Client/CloudJWTProvider.h>
+#include <Common/StringUtils.h>
+#include <Common/ErrorCodes.h>
+
+#include <Poco/Net/HTTPRequest.h>
+#include <Poco/Net/HTTPResponse.h>
+#include <Poco/StreamCopier.h>
+#include <Poco/URI.h>
+#include <Poco/JSON/Parser.h>
+#include <Poco/JSON/Object.h>
+#include <Poco/Dynamic/Var.h>
+#include <Poco/Net/Context.h>
+#include <Poco/Net/HTTPSClientSession.h>
+#include <Poco/Net/SSLManager.h>
+
+#include <jwt-cpp/jwt.h>
+
+#include <thread>
+#include <chrono>
+#if defined(OS_DARWIN) || defined(OS_LINUX)
+#include <spawn.h>
+#include <sys/wait.h>
+#elif defined(OS_WINDOWS)
+#include <windows.h>
+#include <shellapi.h>
+#endif
+#include <iostream>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int SUPPORT_IS_DISABLED;
+    extern const int BAD_ARGUMENTS;
+    extern const int NETWORK_ERROR;
+    extern const int AUTHENTICATION_FAILED;
+    extern const int TIMEOUT_EXCEEDED;
+}
+
+JWTProvider::JWTProvider(
+    std::string auth_url,
+    std::string client_id,
+    std::ostream & out,
+    std::ostream & err)
+    : auth_url_str(std::move(auth_url)),
+      client_id_str(std::move(client_id)),
+      output_stream(out),
+      error_stream(err)
+{
+}
+
+std::string JWTProvider::getJWT()
+{
+    Poco::Timestamp now;
+    Poco::Timestamp expiration_buffer = 15 * Poco::Timespan::SECONDS;
+
+    if (!idp_access_token.empty() && now < idp_access_token_expires_at - expiration_buffer)
+        return idp_access_token;
+
+    if (!idp_refresh_token.empty())
+    {
+        refreshIdPAccessToken();
+        return idp_access_token;
+    }
+
+    deviceCodeLogin();
+    return idp_access_token;
+}
+
+void JWTProvider::deviceCodeLogin()
+{
+    if (client_id_str.empty() || auth_url_str.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error: --auth-client-id and --auth-url are required for --login.");
+
+    Poco::URI device_code_url(auth_url_str + "/oauth/device/code");
+    Poco::URI token_url(auth_url_str + "/oauth/token");
+
+    std::string scope = "openid profile email offline_access";
+    std::string audience = "control-plane-web";
```

Comment:
@kcmannem we should change this to something like `control-plane-cli`

Reply:
or 'sql-console-cli'

Reply:
updated to `token-exchange`